inc/class-previews.php

```diff
@@ -0,0 +1,151 @@
+		remove_action( 'init', '_show_post_preview' );
+
+		// Filter to enable/disable public previews.
+		if ( apply_filters( 'wp_irving_enable_public_previews', false ) ) {
+			add_action( 'wp_irving_components_request', [ $this, 'enabled_public_previews' ] );
+		}
+
+		// After plugins have loaded, check for JWT and modify the preview logic as needed.
+		add_action(
+			'plugins_loaded',
+			function() {
+
+				// Validate that JWT exists and is enabled.
+				if ( ! defined( 'JWT_AUTH_VERSION' ) ) {
+					return;
+				}
+
+				// Hook into Irving's query string to WP_Query process to modify some
+				// logic for previews.
+				add_action( 'wp_irving_components_wp_query', [ $this, 'modify_wp_query_for_previews' ], 10, 4 );
+			}
+		);
+	}
+
+	/**
+	 * Allow-list some query vars for previews.
+	 *
+	 * @param  array $vars Allow-listed query vars.
+	 * @return array
+	 */
+	public function modify_query_vars( array $vars ): array {
+		$vars[] = 'preview_id';
+		$vars[] = 'preview_nonce';
+		return $vars;
+	}
+
+	/**
+	 * Hook into the component API request and change the draft preview to be
+	 * `publish`, mimicking a public post.
+	 *
+	 * @param \WP_REST_Request $request Request object.
+	 */
+	public function enabled_public_previews( \WP_REST_Request $request ) {
+
+		// Require `preview` to be true, and `p` to be an integer.
+		if (
+			false === wp_validate_boolean( $request->get_param( 'preview' ) )
+			|| 0 === absint( $request->get_param( 'p' ) )
+		) {
+			return;
+		}
+
+		add_filter(
+			'posts_results',
+			function ( $posts ) {
+
+				if ( empty( $posts ) ) {
+					return $posts;
+				}
+
+				$posts[0]->post_status = 'publish';
+				return $posts;
+			}
+		);
+	}
+
+	/**
+	 * Modify the \WP_Query object that Irving creates to swap out the published post, for the correct revision.
+	 *
+	 * @param \WP_Query $wp_query      WP_Query object corresponding to this
+	 *                                 request.
+	 * @param string    $path          Request path.
+	 * @param string    $custom_params Custom params.
+	 * @param string    $params        Request params.
+	 * @return \WP_Qury $wp_query WP_Query object for the request.
+	 */
+	public function modify_wp_query_for_previews( $wp_query, $path, $custom_params, $params ) {
+
+		// Auth and a real route required.
+		if ( ! is_user_logged_in() || ! $wp_query->have_posts() ) {
+			return $wp_query;
+		}
+
+		$is_preview = wp_validate_boolean( $params['preview'] ?? false );
+		$preview_id = absint( $params['preview_id'] ?? 0 );
+
+		// Require `preview` to be true, and `preview_id` to be an integer.
+		if ( false === $is_preview || 0 === $preview_id ) {
+			return $wp_query;
+		}
+
+		// Get all the revisions.
+		$revisions = wp_get_post_revisions( $wp_query->post->ID );
+
+		// Use the preview_id, or fallback to the most recent revision.
+		$revision = $revisions[ $preview_id ] ?? current( $revisions );
+
+		$wp_query->posts[0] = $revision;
+		$wp_query->post     = $revision;
```

Comment:
Actually, instead of whole-sale replacing the post, I think we should only replace the title, content, and excerpt, since those are the only things that are allow-listed as revision data in WP Core. In testing, I'm seeing meta data being dropped b/c it's associated with the original published post, and replacing the post object breaks this context.